### PR TITLE
A new AudioEngine sink for sndio

### PR DESCRIPTION
**Pull request by ghost**
## Description
A new AudioEngine sink for [sndio](http://www.sndio.org).

## Motivation and Context

Sndio is an audio framework developed by the OpenBSD project.  So far
it's supports Linux, DragonFly BSD, FreeBSD, and OpenBSD.

I'm currently running a patched version of the Kodi FreeBSD port with
this sink enabled.  It would be nice if this can be included properly.

The pull request includes some changes (mostly adding missing
includes) that were required for building Kodi on FreeBSD.  The ALSA
backend doesn't currently compile on FreeBSD.

## How Has This Been Tested?

I've tested the sink on FreeBSD HEAD and Fedora 25 with Kodi compiled
from the master branch.  I've been patching the FreeBSD port with
versions of this sink for the last 6 months or so.  Ubuntu/Debian have
an sndio package, but on Fedora you have to install it manually.

To test it on Linux you probably need to disable PulseAudio
temporarily to free the audio device.  `sndiod` will lock it while it
is running.  Start `sndiod -dd` so that it runs in the foreground and
prints debug messages.  Kodi was started with `AE_SINK=sndio
./kodi.bin -p` from the build directory.

## Types of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/filesystem/SMBDirectory.cpp [hunk 1]
```diff
@@ -31,6 +31,7 @@
 
 #include "system.h"
 
+#ifdef HAS_FILESYSTEM_SMB
```
Comment (stefansaraev) [inline xbmc/filesystem/SMBDirectory.cpp]: building with -DENABLE_SMBCLIENT=OFF should be enough. see https://github.com/xbmc/xbmc/blob/master/xbmc/filesystem/CMakeLists.txt#L160-L165

File: xbmc/cores/AudioEngine/Sinks/AESinkSNDIO.h [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  static void EnumerateDevicesEx(AEDeviceInfoList &list, bool force = false);
+private:
+  void AudioFormatToPar(AEAudioFormat& format);
+  bool ParToAudioFormat(AEAudioFormat& format);
+  static void OnmoveCb(void *arg, int delta);
+
+  struct sio_hdl *m_hdl;
+  struct sio_par par;
+  ssize_t played;
+  ssize_t written;
```
Comment (fritsch) [inline xbmc/cores/AudioEngine/Sinks/AESinkSNDIO.h]: please use ```m_ ``` for the members

**Comment by fritsch** [comment]
@t6 please rebase and remove the commits not needed.

One thing to add: there is a currently work going on in order to remove the ugly Factory that instantiates the sink (you cannot know that). So if we would merge that one, will you help us to also port this specific sink to the upcoming factory changes?

And something also clear: You will become the maintainer of that one after review is finished, right?

And last but not least did you run time test it?

**Comment by fritsch** [comment]
Also please drop the first commit - the build fixes. This should go in separately after reviewed by @wsnipex - if it needs to go in prior to the sink commit, then so be it.

In general for kodi we make "topic" PRs, e.g. adding one feature, doing exactly one fix and so on.

So in your case:

PR 1: Build fixes
PR 2: AESinkSNDIO Implement new sink for openbsd

Thanks in advance.

**Comment by fritsch** [comment]
One thing I have forgotten: 
It seems that https://github.com/t6/xbmc/blob/2ad118e791a181ad5261f3e052b176ce6ceb75b5/xbmc/cores/AudioEngine/AESinkFactory.cpp#L136 prioritizes sndio over alsa and therefore linux ALSA users might run into issues if they - by accident have sndio-dev dependencies installed and don't take care when building. Please change it that way that ALSA has priority.

**Comment by t6** [comment]
> So if we would merge that one, will you help us to also port this specific sink to the upcoming factory changes?

Sure.

> You will become the maintainer of that one after review is finished, right?

Yes, of course.

> And last but not least did you run time test it?

Yes.

> In general for kodi we make "topic" PRs, e.g. adding one feature, doing exactly one fix and so on.

Sure, I'll open a second PR.  It needs to go in after the sink though.  Probably best to do it after this is in so I can properly rebase it on top of master.

https://github.com/t6/xbmc/compare/patch-sndio-cmake...t6:patch-freebsd

> Please change it that way that ALSA has priority.

Done.

**Comment by MartijnKaijser** [comment]
ping @fritsch

**Comment by AchimTuran** [comment]
@fritsch is this PR ready for a merge?

**Comment by AchimTuran** [comment]
In my view the changes in the sink factory are looking good. Furthermore the sink should not break other platforms as it is not build there.

Jenkins build this please.

**Comment by AchimTuran** [comment]
jenkins build this please

**Comment by AchimTuran** [comment]
@t6
So far I see it was your first PR against our project, right? Thank you very much for your work and contribution. We hope that we can work on more stuff if you want.

**Pull request merged**